packages/grid/_modules_/grid/hooks/features/rows/useGridRowsHydration.ts

```diff
@@ -0,0 +1,109 @@
+import * as React from 'react';
+import { useGridApiEventHandler } from '../../utils/useGridApiEventHandler';
+import { GridEvents } from '../../../models/events/gridEvents';
+import { useGridState } from '../../utils/useGridState';
+import { GridApiRef } from '../../../models/api/gridApiRef';
+import { GridRowsHydrationApi } from '../../../models/api/gridRowsHydrationApi';
+import { GridComponentProps } from '../../../GridComponentProps';
+import { getCurrentPageRows } from '../../utils/useCurrentPageRows';
+import { GridPreProcessingGroup } from '../../core/preProcessing';
+import { GridEventListener } from '../../../models/events/gridEventListener';
+import { useGridApiMethod } from '../../utils/useGridApiMethod';
+import { GridRowId } from '../../../models/gridRows';
+import {
+  GridRowHeightParams,
+  GridRowHeightReturnValue,
+} from '../../../models/params/gridRowParams';
+
+/**
+ * @requires useGridRowGroupsPreProcessing (method)
+ * @requires useGridPageSize (method)
+ * @requires useGridPage (method)
+ */
+export const useGridRowsHydration = (
+  apiRef: GridApiRef,
+  props: Pick<GridComponentProps, 'getRowHeight' | 'pagination' | 'paginationMode'>,
+): void => {
+  const [gridState, setGridState, forceUpdate] = useGridState(apiRef);
+  const rowsHeightCollection = React.useRef(new Map<GridRowId, number>());
+
+  const hydrateRowsMeta = React.useCallback(
+    (getRowHeight?: (params: GridRowHeightParams) => GridRowHeightReturnValue) => {
+      const { rows } = getCurrentPageRows(apiRef.current.state, {
+        pagination: props.pagination,
+        paginationMode: props.paginationMode,
+      });
+
+      setGridState((state) => {
+        const positions: number[] = [];
+        const totalHeight = rows.reduce((acc: number, row) => {
+          positions.push(acc);
+          let targetRowHeight = state.density?.rowHeight;
+
+          if (getRowHeight) {
+            // Default back to base rowHeight if getRowHeight returns null or undefined.
+            targetRowHeight =
+              getRowHeight({ ...row.model, densityFactor: state.density?.factor }) ||
```

Comment:
I would advocate against spreading the model like that.
I would prefer the following:

```suggestion
              getRowHeight({ ...row, densityFactor: state.density?.factor }) ||
```

And 

```ts
export interface GridRowHeightParams extends GridRowEntry {
  densityFactor: number;
}
```